docking/pose_dock.py

```diff
@@ -15,130 +16,183 @@
+    host_pdbfile: path to host pdb file to dock into
+    outdir: where to write output (will be created if it does not already exist)
+    n_steps: how many total steps of simulation to do (recommended: 20000)
+    lowering_steps: how many steps to lower the guest over (recommended: 10000)
+    start_percent: how visible the guest should start out (recommended: 0.75)
+    random_rotation: whether to apply a random rotation to each guest before beginning the simulation
+    constant_atoms: atom numbers from the host_pdbfile to hold mostly fixed across the simulation
+
+    Output
+    ------
+
+    A pdb file every 1000 steps (outdir/<guest_name>_<step>.pdb)
+    stdout every 1000 steps noting the step number, lambda value, and energy
+    """
+    if not os.path.exists(args.outdir):
+        os.makedirs(args.outdir)
+
+    suppl = Chem.SDMolSupplier(guests_sdfile, removeHs=False)
     for guest_mol in suppl:
-        break
+        guest_name = guest_mol.GetProp("_Name")
+        host_mol = Chem.MolFromPDBFile(host_pdbfile, removeHs=False)
+        combined_pdb = Chem.CombineMols(host_mol, guest_mol)
+
+        guest_ff_handlers = deserialize_handlers(
+            open(
+                os.path.join(
+                    os.path.dirname(os.path.abspath(__file__)),
+                    "..",
+                    "ff/params/smirnoff_1_1_0_ccc.py",
+                )
+            ).read()
+        )
+        amber_ff = app.ForceField("amber99sbildn.xml", "tip3p.xml")
+        host_file = PDBFile(host_pdbfile)
+        host_system = amber_ff.createSystem(
+            host_file.topology,
+            nonbondedMethod=app.NoCutoff,
+            constraints=None,
+            rigidWater=False,
+        )
+
+        bps, masses = dock_setup.combine_potentials(
+            guest_ff_handlers, guest_mol, host_system, np.float32, constant_atoms
+        )
+
+        host_conf = []
+        for x, y, z in host_file.positions:
+            host_conf.append([to_md_units(x), to_md_units(y), to_md_units(z)])
+        host_conf = np.array(host_conf)
+        conformer = guest_mol.GetConformer(0)
+        mol_conf = np.array(conformer.GetPositions(), dtype=np.float64)
+        mol_conf = mol_conf / 10  # convert to md_units
+
+        if random_rotation:
+            center = np.mean(mol_conf, axis=0)
+            mol_conf -= center
+            from scipy.stats import special_ortho_group
+
+            mol_conf = np.matmul(mol_conf, special_ortho_group.rvs(3))
+            mol_conf += center
+
+        x0 = np.concatenate([host_conf, mol_conf])  # combined geometry
+        v0 = np.zeros_like(x0)
+
+        seed = 2020
+        intg = LangevinIntegrator(300, 1.5e-3, 1.0, masses, seed).impl()
+
+        box = np.eye(3) * 100
+        v0 = np.zeros_like(x0)
+
+        impls = []
+        for b in bps:
+            p_impl = b.bound_impl()
+            impls.append(p_impl)
+
+        ctxt = custom_ops.Context(x0, v0, box, intg, impls)
+
```

Comment:
Should probably add a sanity assertion that n_steps >= lower_steps, otherwise np.zeros(-1) will throw a cryptic error message:

```
>>> np.zeros(-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: negative dimensions are not allowed
```